src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.CompilationTrackerState.cs

```diff
@@ -99,71 +135,95 @@ public WithCompilationTrackerState(Compilation compilationWithoutGeneratedDocume
                 }
             }
 
+            /// <summary>
+            /// Root type for all compilation tracker that have a compilation, but are not the final <see
+            /// cref="FinalCompilationTrackerState"/>
+            /// </summary>
+            private abstract class NonFinalWithCompilationTrackerState : WithCompilationTrackerState
+            {
+                /// <summary>
+                /// The result of taking the original completed compilation that had generated documents and updating
+                /// them by apply the <see cref="CompilationAndGeneratorDriverTranslationAction" />; this is not a
+                /// correct snapshot in that the generators have not been rerun, but may be reusable if the generators
+                /// are later found to give the same output.
+                /// </summary>
+                public Compilation? StaleCompilationWithGeneratedDocuments { get; }
+
+                protected NonFinalWithCompilationTrackerState(
+                    bool isFrozen,
+                    Compilation compilationWithoutGeneratedDocuments,
+                    CompilationTrackerGeneratorInfo generatorInfo,
+                    Compilation? staleCompilationWithGeneratedDocuments)
+                    : base(isFrozen, compilationWithoutGeneratedDocuments, generatorInfo)
+                {
+                    // If we're not in the frozen state, the documents must *not* be final.  We're not a FinalState, and
+                    // as such, we must still determine what the up to date generated docs are.
+                    Contract.ThrowIfTrue(!IsFrozen && generatorInfo.DocumentsAreFinal);
```

Comment:
Or even stronger:

```suggestion
                    Contract.ThrowIfFalse(IsFrozen == generatorInfo.DocumentsAreFinal)
```

Because if the base class also has an assert about the !IsFrozen case. Since both cases are covered here we can simplify this.

And this also makes sense: in the final state, IsFrozen can go either way. But at all other points, the only way we won't compute further is if we're frozen.

Reply:
yup.  i took the latter.  good call.